crates/polars-parquet/src/arrow/write/pages.rs

```diff
@@ -268,42 +269,133 @@ fn to_nested_recursive(
     Ok(())
 }
 
-/// Convert [`Array`] to `Vec<&dyn Array>` leaves in DFS order.
-pub fn to_leaves(array: &dyn Array) -> Vec<&dyn Array> {
-    let mut leaves = vec![];
-    to_leaves_recursive(array, &mut leaves);
-    leaves
+fn expand_list_validity<'a, O: Offset>(
+    array: &'a ListArray<O>,
+    validity: Option<Bitmap>,
+    arrays: &mut VecDeque<(&'a dyn Array, Option<Bitmap>)>,
+) {
+    let Some(list_validity) = validity else {
```

Comment:
Can we check on null count here. There might be a validity even if all values are valid. This way we can early escape for those cases as well.